plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSnapshotPageSource.java

```diff
@@ -0,0 +1,224 @@
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.internal.schema.InternalSchema;
+import org.apache.hudi.storage.HoodieStorage;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.plugin.hudi.HudiUtil.constructSchema;
+import static org.apache.hudi.common.model.HoodieRecord.HOODIE_META_COLUMNS;
+
+public class HudiSnapshotPageSource
+        implements ConnectorPageSource {
+    private final HoodieStorage storage;
+    private final String basePath;
+    private final HudiSplit split;
+    private final Optional<ConnectorPageSource> baseFilePageSource;
+    private final List<HiveColumnHandle> columnHandles;
+    private final Schema readerSchema;
+    private final TypedProperties payloadProps = new TypedProperties();
+    private final PageBuilder pageBuilder;
+    private final HudiAvroSerializer avroSerializer;
+    private final int recordKeyFieldPos;
+    private final Map<Integer, String> partitionValueMap;
+
+    private Map<String, HoodieRecord> logRecordMap;
+
+    public HudiSnapshotPageSource(List<HivePartitionKey> partitionKeyList,
+                                  HoodieStorage storage,
+                                  String basePath,
+                                  HudiSplit split,
+                                  Optional<ConnectorPageSource> baseFilePageSource,
+                                  List<HiveColumnHandle> dataHandles,
+                                  List<HiveColumnHandle> columnHandles,
+                                  Optional<String> preCombineField) {
+        this.storage = storage;
+        this.basePath = basePath;
+        this.split = split;
+        this.baseFilePageSource = baseFilePageSource;
+        this.columnHandles = columnHandles;
+        this.readerSchema = constructSchema(columnHandles.stream().map(HiveColumnHandle::getName).toList(),
+                columnHandles.stream().map(HiveColumnHandle::getHiveType).toList(), false);
+        this.pageBuilder = new PageBuilder(dataHandles.stream().map(HiveColumnHandle::getType).toList());
+        Map<String, String> nameToPartitionValueMap = partitionKeyList.stream().collect(
+                Collectors.toMap(e -> e.name(), e -> e.value()));
+        this.partitionValueMap = new HashMap<>();
+        for (int i = 0; i < dataHandles.size(); i++) {
+            HiveColumnHandle handle = dataHandles.get(i);
+            if (handle.isPartitionKey()) {
+                partitionValueMap.put(i + HOODIE_META_COLUMNS.size(), nameToPartitionValueMap.get(handle.getName()));
+            }
+        }
+        this.avroSerializer = new HudiAvroSerializer(columnHandles);
+        this.recordKeyFieldPos = 2;
```

Comment:
should we check by name instead of ordinal? Btw, i think we can use `HOODIE_META_COLUMNS_NAME_TO_POS` from `HoodieRecord` instad of hardcoding.

Reply:
Fixed.  The hardcoding is for POC only.